ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/tracing/diff/StateDiffGenerator.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package org.hyperledger.besu.ethereum.api.jsonrpc.internal.results.tracing.diff;
+
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.processor.TransactionTrace;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.results.tracing.Trace;
+import org.hyperledger.besu.ethereum.core.AbstractWorldUpdater;
+import org.hyperledger.besu.ethereum.core.AbstractWorldUpdater.UpdateTrackingAccount;
+import org.hyperledger.besu.ethereum.core.Account;
+import org.hyperledger.besu.ethereum.core.Address;
+import org.hyperledger.besu.ethereum.core.Wei;
+import org.hyperledger.besu.ethereum.core.WorldUpdater;
+import org.hyperledger.besu.ethereum.debug.TraceFrame;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.TreeMap;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import org.apache.tuweni.units.bigints.UInt256;
+
+public class StateDiffGenerator {
+
+  public Stream<Trace> generateStateDiff(final TransactionTrace transactionTrace) {
+    List<TraceFrame> traceFrames = transactionTrace.getTraceFrames();
+    if (traceFrames.size() < 1) {
+      return Stream.empty();
+    }
+
+    // This corresponds to the world state after the TX executed
+    // It is two deep because of the way we addressed Spurious Dragon.
+    WorldUpdater transactionUpdater =
+        traceFrames
+            .get(0)
+            .getMessageFrame()
+            .getWorldState()
+            .parentUpdater()
+            .get()
+            .parentUpdater()
+            .get();
+    // This corresponds to the world state prior to the TX execution,
+    // Either the initial block state or the state of the prior TX
+    WorldUpdater previousUpdater = transactionUpdater.parentUpdater().get();
+
+    StateDiffTrace stateDiffResult = new StateDiffTrace();
+
+    for (Account touchedAccount : transactionUpdater.getTouchedAccounts()) {
+      if (!(touchedAccount instanceof AbstractWorldUpdater.UpdateTrackingAccount)) {
```

Comment:
Why not  filtering the `transactionUpdater.getTouchedAccounts` Collection to have only `UpdateTrackingAccount` types ?